src/frequenz/sdk/microgrid/component_data.py

```diff
@@ -6,240 +6,243 @@
 License
 MIT
 """
+from __future__ import annotations
 
-from abc import ABC
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field
 from datetime import datetime
-from typing import Tuple
+from typing import Optional, Tuple
 
 import frequenz.api.microgrid.microgrid_pb2 as microgrid_pb
 import pytz
 
 from .component_states import EVChargerCableState
 
 
-class _BaseComponentData(ABC):
-    """A private base class for strongly typed component data classes."""
+@dataclass(frozen=True)
+class ComponentData(ABC):
+    """A private base class for strongly typed component data classes.
 
-    _raw: microgrid_pb.ComponentData
+    Attributes:
+        component_id: the ID identifying this component in the microgrid
+        timestamp: the timestamp of when the data was measured.
+    """
 
-    @property
-    def component_id(self) -> int:
-        """Return the component id of the message.
+    component_id: int
+    timestamp: datetime
+    _raw: Optional[microgrid_pb.ComponentData] = field(default=None, init=False)
 
-        Returns:
-            Component id of the message.
-        """
-        return self._raw.id
-
-    @property
-    def timestamp(self) -> datetime:
-        """Return the timestamp of the message.
+    def _set_raw(self, raw: microgrid_pb.ComponentData) -> None:
```

Comment:
This should stay protected, as we don't want users calling this themselves. For the users `raw` is still read-only.

Reply:
I know that we don't want users to call this method, but what do you mean by `protected`? We cannot really make this method inaccessible to the users. The only thing we can do is to make it a private method, which it already is.

Reply:
just using a  leading `_`, that's the convention in python. :)

I was just clarifying that even if `raw` should drop the `_`, this method shouldn't.